url6544f14c72_charles_d_burton_java_websocket_example_tgz-pJ8-minewebsocket_MineWebSocketJ8/src/minewebsocket/org/java_websocket/WebSocketImpl.java

```diff
@@ -0,0 +1,737 @@
+
+	private Role role;
+
+	private Opcode current_continuous_frame_opcode = null;
+
+	/** the bytes of an incomplete received handshake */
+	private ByteBuffer tmpHandshakeBytes = ByteBuffer.allocate( 0 );
+
+	/** stores the handshake sent by this websocket ( Role.CLIENT only ) */
+	private ClientHandshake handshakerequest = null;
+
+	private String closemessage = null;
+	private Integer closecode = null;
+	private Boolean closedremotely = null;
+	
+	private String resourceDescriptor = null;
+
+	/**
+	 * crates a websocket with server role
+	 */
+	public WebSocketImpl( WebSocketListener listener , List<Draft> drafts ) {
+		this( listener, (Draft) null );
+		this.role = Role.SERVER;
+		// draft.copyInstance will be called when the draft is first needed
+		if( drafts == null || drafts.isEmpty() ) {
+			knownDrafts = defaultdraftlist;
+		} else {
+			knownDrafts = drafts;
+		}
+	}
+
+	/**
+	 * crates a websocket with client role
+	 * 
+	 * @param socket
+	 *            may be unbound
+	 */
+	public WebSocketImpl( WebSocketListener listener , Draft draft ) {
+		if( listener == null || ( draft == null && role == Role.SERVER ) )// socket can be null because we want do be able to create the object without already having a bound channel
+			throw new IllegalArgumentException( "parameters must not be null" );
+		this.outQueue = new LinkedBlockingQueue<ByteBuffer>();
+		inQueue = new LinkedBlockingQueue<ByteBuffer>();
+		this.wsl = listener;
+		this.role = Role.CLIENT;
+		if( draft != null )
+			this.draft = draft.copyInstance();
+	}
+
+	@Deprecated
+	public WebSocketImpl( WebSocketListener listener , Draft draft , Socket socket ) {
+		this( listener, draft );
+	}
+
+	@Deprecated
+	public WebSocketImpl( WebSocketListener listener , List<Draft> drafts , Socket socket ) {
+		this( listener, drafts );
+	}
+
+	/**
+	 * 
+	 */
+	public void decode( ByteBuffer socketBuffer ) {
+		assert ( socketBuffer.hasRemaining() );
+
+		if( DEBUG )
+			System.out.println( "process(" + socketBuffer.remaining() + "): {" + ( socketBuffer.remaining() > 1000 ? "too big to display" : new String( socketBuffer.array(), socketBuffer.position(), socketBuffer.remaining() ) ) + "}" );
+
+		if( readystate != READYSTATE.NOT_YET_CONNECTED ) {
+			decodeFrames( socketBuffer );;
+		} else {
+			if( decodeHandshake( socketBuffer ) ) {
+				assert ( tmpHandshakeBytes.hasRemaining() != socketBuffer.hasRemaining() || !socketBuffer.hasRemaining() ); // the buffers will never have remaining bytes at the same time
+
+				if( socketBuffer.hasRemaining() ) {
+					decodeFrames( socketBuffer );
+				} else if( tmpHandshakeBytes.hasRemaining() ) {
+					decodeFrames( tmpHandshakeBytes );
+				}
+			}
+		}
+		assert ( isClosing() || isFlushAndClose() || !socketBuffer.hasRemaining() );
+	}
+	/**
+	 * Returns whether the handshake phase has is completed.
+	 * In case of a broken handshake this will be never the case.
+	 **/
+	private boolean decodeHandshake( ByteBuffer socketBufferNew ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 26. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.